runtime/src/main/java/org/corfudb/runtime/clients/LayoutHandler.java

```diff
@@ -52,46 +59,116 @@ public LayoutClient getClient(long epoch, UUID clusterID) {
+    /**
+     * Handle a propose layout response from the server.
+     *
+     * @param msg The propose layout response message.
+     * @param ctx The context the message was sent under.
+     * @param r A reference to the router.
+     * @return True if ACK, throw an {@link OutrankedException} if REJECT.
+     */
+    @ResponseHandler(type = PayloadCase.PROPOSE_LAYOUT_RESPONSE)
+    private static Object handleProposeLayoutResponse(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                      IClientRouter r) {
+        ProposeLayoutResponseMsg proposeLayoutMsg = msg.getPayload().getProposeLayoutResponse();
+        boolean proposed = proposeLayoutMsg.getProposed();
+        long rank = proposeLayoutMsg.getRank();
+
+        if (!proposed) {
+            throw new OutrankedException(rank);
+        }
+        return true;
     }
 
-    @ClientHandler(type = CorfuMsgType.LAYOUT_PROPOSE_REJECT)
-    private static Object handleProposeReject(CorfuPayloadMsg<LayoutProposeResponse> msg,
-                                              ChannelHandlerContext ctx, IClientRouter r)
-            throws Exception {
-        LayoutProposeResponse response = msg.getPayload();
-        throw new OutrankedException(response.getRank());
+    /**
+     * Handle a propose layout response from the server.
+     *
+     * @param msg The propose layout response message.
+     * @param ctx The context the message was sent under.
+     * @param r A reference to the router.
+     * @return True if ACK, throw an {@link OutrankedException} if REJECT.
+     */
+    @ResponseHandler(type = PayloadCase.COMMIT_LAYOUT_RESPONSE)
+    private static Object handleCommitLayoutResponse(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                     IClientRouter r) {
+        CommitLayoutResponseMsg commitLayoutMsg = msg.getPayload().getCommitLayoutResponse();
+
+        return commitLayoutMsg.getCommitted();
     }
 
-    @ClientHandler(type = CorfuMsgType.LAYOUT_ALREADY_BOOTSTRAP)
-    private static Object handleAlreadyBootstrap(CorfuMsg msg,
-                                                 ChannelHandlerContext ctx, IClientRouter r)
-            throws Exception {
-        throw new AlreadyBootstrappedException();
+    /**
+     * Handle a bootstrap layout response from the server.
+     *
+     * @param msg The propose layout response message.
+     * @param ctx The context the message was sent under.
+     * @param r A reference to the router.
+     * @return True if ACK, false if NACK.
+     */
+    @ResponseHandler(type = PayloadCase.BOOTSTRAP_LAYOUT_RESPONSE)
+    private static Object handleBootstrapLayoutResponse(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                        IClientRouter r) {
+        BootstrapLayoutResponseMsg bootstrapLayoutMsg =  msg.getPayload().getBootstrapLayoutResponse();
+
+        return bootstrapLayoutMsg.getBootstrapped();
```

Comment:
This is a bug. The clients are expecting an exception to be thrown when this response is sent back, but this will return a value.

Reply:
If you look here: https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/view/LayoutView.java#L328

A lot of code paths that rely on the exception to be thrown won't execute, because this changes the semantics of the method, but code that consumes this wasn't changed.

Reply:
Ops, I'm wrong. 

There's a test for this and it is handled in the base handler. Kind of confusing when some errors return a false bool and some throw an exception.